[JENKINS-26687] Better exception message for JSONSignatureValidator.verifySignature()

cyrille-leclerc opened the pull request:
[JENKINS-26687](https://issues.jenkins-ci.org/browse/JENKINS-26687)

Add the SubjectDN of the evaluated X509 certificates when the validation of the JSON file of a "Downloadable Tool Installer" fails.

cyrille-leclerc committed:
[JENKINS-26687] Add the SubjectDN of the evaluated X509 certificates when the validation of the JSON file of a "Downloadable Tool Installer" fails.

olivergondza commented on core/src/main/java/jenkins/util/JSONSignatureValidator.java:
```diff
@@ -73,7 +75,17 @@ public FormValidation verifySignature(JSONObject o) throws IOException {
-                CertificateUtil.validatePath(certs, loadTrustAnchors(cf));
+                try {
+                    CertificateUtil.validatePath(certs, loadTrustAnchors(cf));
+                } catch (GeneralSecurityException | IOException e) {
```
This is java 7.

cyrille-leclerc replied:
You are right, I don't know why I overwritted the java.level value to 1.7 (https://github.com/jenkinsci/jenkins/blob/master/pom.xml#L616 )

jtnord commented on core/src/main/java/jenkins/util/JSONSignatureValidator.java:
```diff
@@ -73,7 +75,17 @@ public FormValidation verifySignature(JSONObject o) throws IOException {
+                    Iterable<Object> certSubjectDNs = Iterables.transform(certs, new Function<X509Certificate, Object>() {
+                        @Override
+                        public Object apply(X509Certificate cert) {
+                            return cert == null ? null : cert.getSubjectDN();
```
rather than null (which will likely print "null" in the warning - would it not be better to use "unknown" or "unknown DN"

jtnord commented on core/src/main/java/jenkins/util/JSONSignatureValidator.java:
```diff
@@ -73,7 +75,17 @@ public FormValidation verifySignature(JSONObject o) throws IOException {
+                            return cert == null ? null : cert.getSubjectDN();
+                        }
+                    });
+                    return FormValidation.warning(e,String.format("Certificate validation failed with certs %s in %s", certSubjectDNs.toString(), name));
```
if a cert is expired will it still validate?  
if not you have lost the not yet valid warning above.

jtnord replied:
The context of the exception is invaluable (will contain the reason validation failed).
It must at least be logged - and I would be tempted to included it in the validation result so that the user sees it.  (this will also fix the loss of information I mentioned above.)

jglick replied:
Right, the `cause` already contains the interesting information:

    NotAfter: Tue Jan 27 15:04:07 CST 2015

which this patch would apparently throw away. :-1:

cyrille-leclerc replied:
I really liked the detailed message (see below and [JENKINS-26688](https://issues.jenkins-ci.org/browse/JENKINS-26688)). I feel it is particularly valuable when different tools are signed by different certificates (CloudBees uses its own certificate to sign its specific Downloadable Tool Installers).

```
Certificate validation failed with certs [CN=Community Update Center, O=Jenkins Project, ST=California, C=US] in downloadable hudson.tools.JDKInstaller java.security.cert.CertPathValidatorException: timestamp check failed
	at sun.security.provider.certpath.PKIXMasterCertPathValidator.validate(PKIXMasterCertPathValidator.java:129)
	...
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.security.cert.CertificateExpiredException: NotAfter: Tue Jan 27 22:04:07 CET 2015
	at sun.security.provider.certpath.PKIXMasterCertPathValidator.validate(PKIXMasterCertPathValidator.java:119)
	... 
```

cyrille-leclerc committed:
Add the SubjectDN of the evaluated X509 certificates when the validation of the JSON file of a "Downloadable Tool Installer" fails.
Error message:

```
Certificate validation failed with certs [CN=Community Update Center, O=Jenkins Project, ST=California, C=US] for downloadable hudson.tools.JDKInstaller: java.security.cert.CertPathValidatorException: timestamp check failed
```

cyrille-leclerc commented:
The exception message is:

```
Certificate validation failed with certs [CN=Community Update Center, O=Jenkins Project, ST=California, C=US] for downloadable hudson.tools.JDKInstaller: java.security.cert.CertPathValidatorException: timestamp check failed
java.security.cert.CertPathValidatorException: timestamp check failed
	at sun.security.provider.certpath.PKIXMasterCertPathValidator.validate(PKIXMasterCertPathValidator.java:129)
	...
Caused by: java.security.cert.CertificateExpiredException: NotAfter: Tue Jan 27 22:04:07 CET 2015
	at sun.security.x509.CertificateValidity.valid(CertificateValidity.java:274)
	...
```

jglick commented:
+0 I guess. The only interesting part here is the `NotAfter: Tue Jan 27 22:04:07 CET 2015`, which is not returned. The function does advertise that it will return a `FormValidation`, which this patch accomplishes.

However it does not catch and display `CertificateExpiredException` specially, like the code above it does. Why `checkValidity` would _not_ throw this when `validatePath` apparently does is mysterious to me. Smells like a mistake in the code that ought to be fixed and covered by a unit test.

jtnord commented on core/src/main/java/jenkins/util/JSONSignatureValidator.java:
```diff
@@ -73,7 +75,17 @@ public FormValidation verifySignature(JSONObject o) throws IOException {
+                            return cert == null ? null : cert.getSubjectDN();
+                        }
+                    });
+                    return FormValidation.warning(e,String.format("Certificate validation failed with certs %s for %s: %s", certSubjectDNs.toString(), name, e.toString()));
```
Isn't one of the root causes here that the earlier warnings should be returned to the user and they where not as validatePath then failed?

(see lines 70,72)
The message as returned is pretty nasty to comprehend and will swallow the more human parsable message from the mentioned lines.

jtnord commented on core/src/main/java/jenkins/util/JSONSignatureValidator.java:
```diff
@@ -73,7 +75,17 @@ public FormValidation verifySignature(JSONObject o) throws IOException {
 
-                CertificateUtil.validatePath(certs, loadTrustAnchors(cf));
+                try {
+                    CertificateUtil.validatePath(certs, loadTrustAnchors(cf));
```
why are we even attempting this if the certrificate is not valid (lines 70,72) above.

Isn't this the root cause the human messages that would have been displayed are swallowed as this code ran and threw an exception?

oleg-nenashev commented:
Marked as WiP

oleg-nenashev commented:
@cyrille-leclerc 
Any plans to finish it?

oleg-nenashev commented:
@cyrille-leclerc bump

cyrille-leclerc commented:
@oleg-nenashev i didn't catch the notification, i'll review

daniel-beck commented:
PR appears to have been abandoned, closing.
